### PR TITLE
Shimlayer: Light up TypeArgumentNullableAnnotations

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Precondition for #8413

Exposes [INamedTypeSymbol.TypeArgumentNullableAnnotations](https://learn.microsoft.com/zh-tw/dotnet/api/microsoft.codeanalysis.inamedtypesymbol.typeargumentnullableannotations?view=roslyn-dotnet-4.7.0) (but not [IMethodSymbol.TypeArgumentNullableAnnotations](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.imethodsymbol.typeargumentnullableannotations?view=roslyn-dotnet-4.7.0))

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs [hunk 1]
```diff
@@ -4,23 +4,89 @@
...
         public static ImmutableArray<IFieldSymbol> TupleElements(this INamedTypeSymbol symbol) => TupleElementsAccessor(symbol);
 
         public static bool IsSerializable(this INamedTypeSymbol symbol) => IsSerializableAccessor(symbol);
+
+        public static ImmutableArray<NullableAnnotation> TypeArgumentNullableAnnotations(this INamedTypeSymbol symbol) => TypeArgumentNullableAnnotationsAccessor(symbol);
+
+        private static Func<INamedTypeSymbol, ImmutableArray<NullableAnnotation>> CreateTypeArgumentNullableAnnotationsAccessor()
+        {
+            // INamedTypeSymbol.TypeArgumentNullableAnnotations is ImmutableArray<Roslyn.NullableAnnotation>
+            // The generated code
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs:36]: Maybe it is far easier to use [`ImmutableArray.CreateRange<TSource,TResult>(ImmutableArray<TSource>, Func<TSource,TResult>)`](https://learn.microsoft.com/en-us/dotnet/api/system.collections.immutable.immutablearray.createrange?view=net-8.0#system-collections-immutable-immutablearray-createrange-2(system-collections-immutable-immutablearray((-0))-system-func((-0-1)))) instead?

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs]: Done in e1a5ef63f534c8f270554868e731d681debb6224

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs [hunk 2]
```diff
@@ -1,26 +1,81 @@
...
+            TypeArgumentNullableAnnotationsAccessor = CreateTypeArgumentNullableAnnotationsAccessor();
         }
 
         public static INamedTypeSymbol TupleUnderlyingType(this INamedTypeSymbol symbol) => TupleUnderlyingTypeAccessor(symbol);
 
         public static ImmutableArray<IFieldSymbol> TupleElements(this INamedTypeSymbol symbol) => TupleElementsAccessor(symbol);
 
         public static bool IsSerializable(this INamedTypeSymbol symbol) => IsSerializableAccessor(symbol);
+
+        public static ImmutableArray<NullableAnnotation> TypeArgumentNullableAnnotations(this INamedTypeSymbol symbol) => TypeArgumentNullableAnnotationsAccessor(symbol);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs:31]: Is this present in the StyleCop repo too, or is it only our addition? 
Shouldn't it follow the new file structure where we said that we'll put our changes on a side?

File: analyzers/tests/SonarAnalyzer.Test/Wrappers/INamedTypeSymbolExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+                    o.ToString();
+                }
+            }
+            """;
+        var (tree, semanticModel) = TestHelper.CompileCS(code);
+        var identifier = tree.GetRoot().DescendantNodes().OfType<IdentifierNameSyntax>().Last(x => x.NameIs("o")); // o in o.ToString()
+        var namedType = semanticModel.GetTypeInfo(identifier).Type.Should().BeAssignableTo<INamedTypeSymbol>().Which;
+        var typeArgumentNullabilityShim = namedType.TypeArgumentNullableAnnotations();
+        var typeArgumentNullability = namedType.TypeArgumentNullableAnnotations;
+        typeArgumentNullabilityShim.Should().BeEquivalentTo([expected]).And.BeEquivalentTo(typeArgumentNullability.Select(x => (NullableAnnotation)x));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.Test/Wrappers/INamedTypeSymbolExtensionsTests.cs:54]: This is heavy and repetitive. Let's extract it.
```suggestion
        ValidateTypeArgumentNullableAnnotations(code, expected);
```

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
One question about the code placement

File: analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ShimLayer/INamedTypeSymbolExtensions.cs]: We defined how to do our extensions of ShimLayer on this card:
https://trello.com/c/fxLbcSr2/77-shimlayer-how-do-we-keep-track-of-our-own-changes?filter=shim
and that resulted in this document
https://docs.google.com/document/d/1hkE_jH8vjsT7phgZ16-RF-uNYmcNDGN8ITBr4pBTGMc/edit#heading=h.iov8d4fboa5j

We went for `Sonar` directory there.

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**